File: sort_beowulf.py
import sys
import time


def n2_sort(lines):
	for i in range(len(lines)-1):
		best = i
		for j in range(i, len(lines)):
			if lines[j].lower() < lines[best].lower():
				best = j
		lines[i], lines[best] = lines[best], lines[i]
	return lines


def nlogn_sort(lines):
	n = len(lines)
	if n <= 1:
		return lines
	# Modified to enhance lowest recursive calls
	elif n == 2:
		if lines[1].lower() > lines[0].lower():
			return [lines[1], lines[0]]
		else:
			return lines
	else:
		half = len(lines) // 2
		return merge(nlogn_sort(lines[:half]), nlogn_sort(lines[half:]))


def merge(A, B):
	S = []
	ai = 0
	bi = 0
	while ai < len(A) and bi < len(B):
		if A[ai].lower() <= B[bi].lower():
			x = A[ai]
			ai += 1
		else:
			x = B[bi]
			bi += 1
		S.append(x)
	return S + A[ai:] + B[bi:]


def main(args):
	if len(args) != 3:
		print('Usage  $ ' + args[0] + ' beowulf.txt' + ' <num_lines>')
		sys.exit(0)
	else:
		filename = args[1]

	num_lines = int(args[2])
	if num_lines > 40707:
		num_lines = 40707
	
	print("Reading n = %d lines" %num_lines)
	lines = []
	with open(filename, 'r') as f:
		for i in range(num_lines):
			lines.append(f.readline().strip())

	print("First 10 words:")
	print(lines[:10])
	#print("Running python merge sort...")
	#print("Running python selection sort...")

	start = time.perf_counter()
	lines = sorted(lines, key=str.lower)
	#lines = nlogn_sort(lines)
	#lines = n2_sort(lines)
	end = time.perf_counter()
	
	print("First 10 words:")
	print(lines[:10])
	print("elapsed time: " + str(end-start) + " seconds")



if __name__ == "__main__":
	main(sys.argv)

File: sort_beowulf.cpp

#include <chrono>
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <cctype>

using namespace std;

// Because for some reason C++ doesn't have case insensitive sorting...
bool str_less(string s1, string s2) {
	int i;
	for (i=0; i < s1.size() && i < s2.size(); i++) {
		if ((s1[i]|32) < (s2[i]|32)) {
			return true;
		} else if ((s1[i]|32) > (s2[i]|32)) {
			return false;
		}
	}
	if (s1.size() < s2.size()) {
		return true;
	}
	return false;
}


vector<string> readtext(char* filename, int num_lines) {
	vector<string> lines;
	string line;
	ifstream infile(filename);

	int i;
	for (i=0; i < num_lines; i++) {
		infile >> line;
		lines.push_back(line);
	}
	return lines;
}


void n2_sort(vector<string> &lines) {
	int i, j, best;
	int max = lines.size();

	for (i = 0; i < max-1; i++) {
		best = i;
		for (j = i+1; j < max; j++) {
			if (str_less(lines[j], lines[best])) {
				best = j;
			}
		}
		swap(lines[i], lines[best]);
	}
}


void print10(vector<string> lines) {
	int i;
	cout << "First 10 words:\n";
	for (i=0; i < 10; i++) {
		cout << "[" << lines[i] << "] ";
	}
	cout << endl;
}


int main(int argc, char** argv) {
	if (argc != 3)
		{
		std::cout << "Usage:  $ ./sort_beowulf beowulf.txt <num_lines>\n";
		return 0;
		}

	int num_lines = stoi(argv[2]);
	if (num_lines > 40707) num_lines = 40707;
	vector<string> lines = readtext(argv[1], num_lines);
	
	cout << "Requested n = " << num_lines << endl;
	print10(lines);
	cout << "Running C++ selection sort...\n";

	auto start = chrono::high_resolution_clock::now();
	n2_sort(lines);
	auto end = chrono::high_resolution_clock::now();

	print10(lines);

	int microseconds = chrono::duration_cast<chrono::microseconds>(end - start).count();
	double seconds = microseconds / 1E6;
	cout << "elapsed time: " << seconds << " seconds" << endl;

	return 0;
}
